cmd/quadlet/main.go

```diff
@@ -0,0 +1,246 @@
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/containers/quadlet/pkg/quadlet"
+	"github.com/containers/quadlet/pkg/systemdparser"
+)
+
+var verboseFlag bool
+var versionFlag bool
+var isUser bool
+
+var noKmsg bool = false
+var kmsgFile *os.File = nil
+
+func logToKmsg(s string) bool {
+	if noKmsg {
+		return false
+	}
+
+	if kmsgFile == nil {
+		f, err := os.OpenFile("/dev/kmsg", os.O_WRONLY, 0644)
+		if err != nil {
+			noKmsg = true
+			return false
+		}
+		kmsgFile = f
+	}
+
+	if _, err := kmsgFile.Write([]byte(s)); err != nil {
+		kmsgFile.Close()
+		kmsgFile = nil
+		return false
+	}
+
+	return true
+}
+
+func Log(format string, a ...any) {
+	s := fmt.Sprintf(format, a...)
+	line := fmt.Sprintf("quadlet-generator[%d]: %s", os.Getpid(), s)
+
+	if !logToKmsg(line) {
+		// If we can't log, print to stderr
+		fmt.Fprintf(os.Stderr, "%s\n", line)
+		os.Stderr.Sync()
+	}
+}
+
+var debugEnabled bool = false
+
+func enableDebug() {
+	debugEnabled = true
+}
+
+func Debug(format string, a ...any) {
+	if debugEnabled {
+		Log(format, a...)
+	}
+}
+
+func getUnitDirs(user bool) []string {
+
+	unitDirsEnv := os.Getenv("QUADLET_UNIT_DIRS")
+	if len(unitDirsEnv) > 0 {
+		return strings.Split(unitDirsEnv, ":")
+	}
+
+	dirs := make([]string, 0)
+	if user {
+		if configDir, err := os.UserConfigDir(); err == nil {
+			dirs = append(dirs, path.Join(configDir, "containers/systemd"))
+		}
+	} else {
+		dirs = append(dirs, quadlet.UnitDirAdmin)
+		dirs = append(dirs, quadlet.UnitDirDistro)
+	}
+	return dirs
+}
+
+func loadUnitsFromDir(sourcePath string, units map[string]*systemdparser.UnitFile) {
+	files, err := os.ReadDir(sourcePath)
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			Log("Can't read \"%s\": %s", sourcePath, err)
```

Comment:
I recommend returning errors and let the callers take care of logging or considering it fatal. Only logging errors quickly turns into spaghetti code.

Applies to other parts of the code as well.

Reply:
What do you mean by callers? This is the main.go file of a systemd generator. There is no caller of this code. And it has to log, directly to /dev/kmsg, because it is running in an extremely limited environment during early boot where there is no other way to communicate to anything at all.

Reply:
I was referring to code hygiene and being more Go idiomatic.

```
if err != nil {
    Log(err)
}
return
```

should be
```
if err != nil {
    return err
}
```

and let the callers of this function do the logging.  It certainly works as is but it is very uncommon in Go - it loves throwing its errors around.

Reply:
Maybe that makes sense for this error, but later in this function we log parse error for each .container file, but then continue loading the others. So, for that case we need to log directly.